Fix #21633: Added screenshot feature for failed acceptance tests

omgupta-iitk opened the pull request:
## Overview

<!--
READ ME FIRST:
Please answer *all* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->
The primary goal of this PR is to upload screenshots of CI test failures for acceptance tests. This will aid in debugging flaky tests and facilitate the implementation of new acceptance tests in the local environment.
1. This PR fixes #21633 
2. This PR does the following: Adds the Screenshot feature for failed acceptance tests and then uploads it as artifacts

## Essential Checklist

Please follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Make-a-pull-request).

- [x] I have linked the issue that this PR fixes in the "Development" section of the sidebar.
- [x] I have checked the "Files Changed" tab and confirmed that the changes are what I want to make.
- [x] I have written tests for my code.
- [x] The **PR title** starts with "Fix #bugnum: " or "Fix part of #bugnum: ...", followed by a short, clear summary of the changes.
- [x] I have assigned the correct reviewers to this PR (or will leave a comment with the phrase "@{{reviewer_username}} PTAL" if I can't assign them directly).

## Testing doc (for PRs with Beam jobs that modify production server data)

<!--
If this PR affects production server data, please follow
[these instructions](https://github.com/oppia/oppia/wiki/Testing-jobs-and-other-features-on-production#submitting-a-pr-with-a-new-job-or-feature-that-requires-third-party-api)
and link to the job request doc here.

Otherwise, please delete this section.
-->

- [ ] A testing doc has been written: ... (ADD LINK) ...
- [ ] _(To be confirmed by the server admin)_ All jobs in this PR have been verified on a live server, and the PR is safe to merge.

## Proof that changes are correct

<!--
Add before-and-after videos/screenshots of the user-facing interface (including
the browser devtools console) to demonstrate that the changes made in this PR
work correctly. Make sure the actions taken in the before and after videos are
the same. (For changes involving responsiveness or adjustment of UI elements,
this should be a video that gradually resizes the viewport from wide to narrow
and back again.) If this PR is for a developer-facing feature, provide
videos/screenshots of the developer-facing interface instead.

When you make updates to the PR, please update these videos/screenshots as well.
You can drop videos/screenshots from previous versions of the PR.

The above should be done for all PRs, including short ones (e.g. a single-line change).
However, if the changes in your PRs are autogenerated via a script and you cannot
provide proof for the changes then please leave a comment "No proof of changes
needed because {{Reason}}" and remove all the sections below.
-->

#### Proof of changes on desktop
 https://github.com/user-attachments/assets/3b69f957-1859-4515-a78b-6123b5cd6107

<!--
Make sure to properly verify that everything works correctly, and that there are
no weird UI mistakes or other problems. Also, if there are any newly added fields,
try to fill them out and test that different inputs are correctly accepted/rejected.

Throttle the network (to 3G) using the browser Developer Tools (see references below).
There should be no performance or UI issues while the network is slow.

References:
 - Chrome: https://css-tricks.com/throttling-the-network/
 - Firefox: https://developer.mozilla.org/en-US/docs/Tools/Network_Monitor/Throttling
-->

#### Proof of changes on mobile phone
https://github.com/user-attachments/assets/6863f302-5ae7-440f-8888-1a5698653872

### Proof of changes, that this actually works. 
 Here a flake error has occurred while submitting the PR: https://github.com/oppia/oppia/actions/runs/13483340920/job/37671539471?pr=21992
the screenshot can be seen Uploaded by the workflow.
(Note: screenshot uploaded on the first error occurrence, not on the 2nd attempt when screen recording is done, so when opening the above GitHub action link, scroll up to the first occurrence of error to validate the claims.)

Artifacts: [desktop-acceptance-test-screenshot-curriculum-admin_create-publish-unpublish-and-delete-topic-and-skill.zip](https://github.com/oppia/oppia/actions/runs/13483340920/artifacts/2637199430)

<!--
In some cases this is not needed (e.g. for pages that we do not expect to
support mobile phones, or for backend-only features).

Feel free to use the Developer Tools emulator for this.

References:
 - Chrome: https://developer.chrome.com/docs/devtools/device-mode/
 - Firefox: https://firefox-source-docs.mozilla.org/devtools-user/index.html#responsive-design-mode
-->


## PR Pointers

- Never force push! If you do, your PR will be closed.
- To reply to reviewers, follow these instructions: https://github.com/oppia/oppia/wiki/Make-a-pull-request#step-5-address-review-comments-until-all-reviewers-approve
- Some e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR).
- See the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia's-code-owners-and-checks-to-be-carried-out-by-developers) for what code owners will expect.

omgupta-iitk committed:
Acceptance Tests, Screenshot feature added

omgupta-iitk committed:
Added the custom-jest-environment file to root files config

omgupta-iitk commented:
The above check fail is a flake and the issue associated with this is : #20988, Thanks.

omgupta-iitk committed:
Fixed Screenshot capturing on the retry attempt

omgupta-iitk commented:
Hi @Vir-8, the PR is ready for review. PTAL at your convenience. Thanks!

omgupta-iitk committed:
Merge branch 'develop' into screenshotAdded

omgupta-iitk committed:
Merge branch 'develop' into screenshotAdded

Vir-8 commented:
Hey @omgupta-iitk, thanks for the PR!

One question, is this supposed to upload the screenshots for CI tests too, or just locally testing them? I see multiple screenshots being uploaded for failing tests in this PR, is that intended?

omgupta-iitk commented:
@Vir-8 Yes, the primary goal of this PR is to upload screenshots of CI test failures for acceptance tests. This will aid in debugging flaky tests and facilitate the implementation of new acceptance tests in the local environment.

Also, this feature can be verified as I encountered a flake in a commit of this PR, and the screenshots got uploaded perfectly; then, in the next commit, I made some changes so that the screenshot wouldn't be captured in the retry attempt where video recording takes place.

Sorry for the inconvenience—I should have included this in the PR description.

omgupta-iitk commented:
@Vir-8 PTAL , thanks.

omgupta-iitk commented:
@seanlip PTAL, Can you tell me if it is okay for us to not take and upload the screenshot during the retry attempt when the video recording is performed? This will upload the screenshots two times and can be messy.

seanlip commented:
@omgupta-iitk No need to update the screenshot for the second retry. Thanks for checking.

omgupta-iitk commented:
@Vir-8 PTAL, thanks!

Vir-8 commented on .github/workflows/full_stack_tests.yml:
```diff
@@ -163,6 +163,12 @@ jobs:
           xvfb-run -a --server-args="-screen 0, 1285x1000x24" 
           python -m scripts.run_acceptance_tests 
           --skip-build --suite=${{ matrix.suite.name }} --prod_env
+      - name: Uploading Desktop test screenshots as Artifacts
```
```suggestion
      - name: Upload desktop test screenshots as artifacts
```

omgupta-iitk replied:
Done

Vir-8 commented on .github/workflows/full_stack_tests.yml:
```diff
@@ -184,6 +190,12 @@ jobs:
           path: /home/runner/work/oppia/oppia/core/tests/test-modules-mappings/acceptance/${{ steps.generate_suite_name_for_artifacts.outputs.MODIFIED_SUITE_NAME }}.txt
       - name: Run Mobile Acceptance Test ${{ matrix.suite.name }}
         run: xvfb-run -a --server-args="-screen 0, 1285x1000x24" python -m scripts.run_acceptance_tests --skip-build --suite=${{ matrix.suite.name }} --prod_env --mobile
+      - name: Uploading Mobile test screenshots as Artifacts
```
```suggestion
      - name: Upload mobile test screenshots as artifacts
```

omgupta-iitk replied:
Done

Vir-8 commented on core/tests/puppeteer-acceptance-tests/custom-jest-environment.js:
```diff
@@ -0,0 +1,40 @@
+// Copyright 2024 The Oppia Authors. All Rights Reserved.
```
2024 -> 2025

omgupta-iitk replied:
Done

Vir-8 commented on core/tests/puppeteer-acceptance-tests/custom-jest-environment.js:
```diff
@@ -0,0 +1,40 @@
+// limitations under the License.
+
+/**
+ * @fileoverview Custom jest environment file.
```
Can we make this comment more descriptive / what this environment is used for etc?

omgupta-iitk replied:
Done

Vir-8 commented on core/tests/puppeteer-acceptance-tests/custom-jest-environment.js:
```diff
@@ -0,0 +1,40 @@
+
+class CustomJestEnvironment extends NodeEnvironment {
+  async handleTestEvent(event) {
+    if (event.name === 'test_done') {
+      if (event.test.errors.length > 0) {
```
```suggestion
    if (event.name === 'test_done' && event.test.errors.length > 0) {
```

omgupta-iitk replied:
Done

Vir-8 reviewed (changes requested):
Mostly LGTM, some minor comments. Thanks for the PR! @omgupta-iitk

omgupta-iitk committed:
Merge branch 'develop' into screenshotAdded

omgupta-iitk committed:
Fixed comments

omgupta-iitk committed:
Fixed lints

omgupta-iitk commented:
@Vir-8 PTAL, thanks!

omgupta-iitk commented:
@Vir-8 PTAL, the test is failing due to a flake #20988

seanlip commented:
@omgupta-iitk If your CI checks fail, please follow [these instructions](https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR), including reporting the flake. Thanks.

omgupta-iitk commented:
@seanlip @Vir-8 Sorry, Actually, this happened twice, and in the previous fail, I added the comment for that.
Added again: https://github.com/oppia/oppia/issues/20988#issuecomment-2692276261
Thanks!

omgupta-iitk committed:
Merge branch 'develop' into screenshotAdded

omgupta-iitk commented:
I accidentally deleted my fork, which caused this PR to close. I’ve recreated it as #22043.
